pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/reducers/SegmentPreprocessingReducer.java

```diff
@@ -0,0 +1,86 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.hadoop.job.reducers;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.mapred.AvroKey;
+import org.apache.avro.mapred.AvroValue;
+import org.apache.avro.mapreduce.AvroMultipleOutputs;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.hadoop.job.JobConfigConstants.*;
+
+
+public class SegmentPreprocessingReducer<T>
+    extends Reducer<T, AvroValue<GenericRecord>, AvroKey<GenericRecord>, NullWritable> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessingReducer.class);
+
+  private AvroMultipleOutputs _multipleOutputs;
+  private AtomicInteger _counter;
+  private int _maxNumberOfRecords;
+  private String _filePrefix;
+
+  @Override
+  public void setup(Context context) {
+    LOGGER.info("Using multiple outputs strategy.");
+    Configuration configuration = context.getConfiguration();
+    _multipleOutputs = new AvroMultipleOutputs(context);
+    _counter = new AtomicInteger();
+    // If it's 0, the output file won't be split into multiple files.
+    // If not, output file will be split when the number of records reaches this number.
+    _maxNumberOfRecords = configuration.getInt(MAXIMUM_NUMBER_OF_RECORDS, 0);
+    LOGGER.info("Maximum number of records per file: {}", _maxNumberOfRecords);
+    _filePrefix = RandomStringUtils.randomAlphanumeric(4);
+  }
+
+  @Override
+  public void reduce(final T inputRecord, final Iterable<AvroValue<GenericRecord>> values, final Context context)
+      throws IOException, InterruptedException {
+    for (final AvroValue<GenericRecord> value : values) {
+      String fileName = generateFileName();
+      _multipleOutputs.write(new AvroKey<>(value.datum()), NullWritable.get(), fileName);
+    }
+  }
+
+  @Override
+  public void cleanup(Context context) throws IOException, InterruptedException {
+    LOGGER.info("Clean up reducer.");
+    if (_multipleOutputs != null) {
+      _multipleOutputs.close();
+      _multipleOutputs = null;
+    }
+    LOGGER.info("Finished cleaning up reducer.");
+  }
+
+  private String generateFileName() {
+    if (_maxNumberOfRecords == 0) {
+      return _filePrefix;
+    } else {
+      return _filePrefix + (_counter.getAndIncrement() / _maxNumberOfRecords);
```

Comment:
as discussed offline, i think we should disable giving people the option to do this resizing for now, as worst case, we will end up with many tiny segments and many big segments (potentially one tiny segment for each reducer  * num of pushes people will be doing)

Reply:
I don't think so. You cannot say it potentially generates too many segments so that it should not be used. At least this is the first phase functionality that users can tune the number of records by themselves. It is users' responsibility to provide a practical config. To enhance this, the next phase can have some ways to mitigate the size of the last file problem.

Reply:
Can we hide the parameter and disallow setting it for now? I think it's a pretty big risk for clients because each reducer can have 1 tiny and the rest giant, and if users use this setting over a long period of time, they will have to completely rebootstrap all their data. They will expect that we will do this work intelligently, but we won't have that until the next phase. In prod, this can be very dangerous - users can have downtime and will to rebootstrap their entire dataset to fix this problem